### PR TITLE
fix duration event of timeline summary

### PR DESCRIPTION
**Pull request by yuanhuihui**
Need to modify duration event to beginEnd event,   because of PR https://dart-review.googlesource.com/c/sdk/+/127921

### CODE REVIEW

**Comment by yuanhuihui** [comment]
relation to https://github.com/flutter/flutter/issues/47771

File: packages/flutter_driver/test/src/timeline_summary_test.dart
Comment (liyuqian) [inline packages/flutter_driver/test/src/timeline_summary_test.dart]: Nit: it seems that we don't have to define `frameBegin` and `frameEnd` separately. We can just reuse the `build` signature, and does the work of `frameBegin` and `frameEnd` inside. That removes the possibility that someone could call multiple `frameBegin` and forget to call any `frameEnd` during a test.

File: packages/flutter_driver/lib/src/driver/timeline_summary.dart
Comment (liyuqian) [inline packages/flutter_driver/lib/src/driver/timeline_summary.dart]: Nit: It seems that we can preserve the signature of `_extractFrameDurations` and just change its implementation. Is there any specific reason that we must have a separate `_extractFrameEvents` function?

**Review by liyuqian — COMMENTED** [review]
The PR itself looks good. I'm more concerned if we can have a Dart change like https://dart-review.googlesource.com/c/sdk/+/127921 but none of our framework tests would fail to notify that such Dart change is in some way breaking the Flutter framework. I also didn't notice any significant change in https://flutter-dashboard.appspot.com/benchmarks.html . @goderbauer : do you know if there's any integration test that's supposed to catch such Dart changes?

**Comment by yuanhuihui** [comment]
> The PR itself looks good. I'm more concerned if we can have a Dart change like https://dart-review.googlesource.com/c/sdk/+/127921 but none of our framework tests would fail to notify that such Dart change is in some way breaking the Flutter framework. I also didn't notice any significant change in https://flutter-dashboard.appspot.com/benchmarks.html . @goderbauer : do you know if there's any integration test that's supposed to catch such Dart changes?


 https://dart-review.googlesource.com/c/sdk/+/127921, it has merged in out own engine. 
 I run command of "flutter drive" to test performance that report build failed.  My PR merged into dart warehouse, not synchronized into engine?

File: packages/flutter_driver/lib/src/driver/timeline_summary.dart
Comment (yuanhuihui) [reply packages/flutter_driver/lib/src/driver/timeline_summary.dart]: changing the implementation is good suggestion. _extractFrameEvents  just keep same style of _extractGpuRasterizerDrawEvents. I have updated PR, remove redundant code and reduced method nested calls

File: packages/flutter_driver/test/src/timeline_summary_test.dart
Comment (yuanhuihui) [reply packages/flutter_driver/test/src/timeline_summary_test.dart]: this test is used to verify that the data can be obtained correctly when the `frameBegin` and `frameEnd` are not matched.  So add test('skips leading "end" events'）and test('skips leading "end" events'), keeping the same with the `begin` and `end`.

**Review by liyuqian — APPROVED** [review]
@yuanhuihui : ah, you're right! Flutter framework is using a 7-day old engine right before your commit has been merged into the Dart SDK Github repo:
- Flutter framework is using engine https://github.com/flutter/engine/commit/42bb7c9eea9a8c165814dc0242e34b276d6c75ce with a commit date of Dec 21, 2019, 7:27 AM GMT+8
- https://dart-review.googlesource.com/c/sdk/+/127921 has been in Github as https://github.com/dart-lang/sdk/commit/006f611a2b9d1b31a2c1e353bf408062e0fa4863 on Dec 21, 2019, 7:37 AM GMT+8

LGTM to the updated PR.

CC @franciscojma86 : I think the recent post-submit failures of the engine roll that you reverted is related to this PR. The engine roll probably should be paused until the landing of this PR.

**Comment by franciscojma86** [comment]
@liyuqian yeah I stopped it yesterday. I was trying to figure out how to fix this so happy to see this PR :)

**Comment by franciscojma86** [comment]
The build is green again @yuanhuihui, is this ok to land?

**Comment by franciscojma86** [comment]
I'll land to unblock the roll. Will revert if needed.

**Comment by franciscojma86** [comment]
This seems to break the build 
```
2019-12-27T09:51:44.476526: stdout: [+9570 ms] 00:10 [32m+0[0m[31m -1[0m: scrolling performance test complex_layout_scroll_perf [1m[31m[E][0m[0m
2019-12-27T09:51:44.477108: stdout: [        ]   Invalid argument(s): durations is empty!
2019-12-27T09:51:44.498305: stdout: [  +21 ms]   package:flutter_driver/src/driver/timeline_summary.dart 167:7   TimelineSummary._averageInMillis
stdout: [        ]   package:flutter_driver/src/driver/timeline_summary.dart 33:12   TimelineSummary.computeAverageFrameBuildTimeMillis
stdout: [        ]   package:flutter_driver/src/driver/timeline_summary.dart 89:42   TimelineSummary.summaryJson
stdout: [        ]   package:flutter_driver/src/driver/timeline_summary.dart 130:4
```
Will revert for now

**Comment by flar** [comment]
This PR needs to include a new hash in `bin/internal/engine.version` to pull over the appropriately patched Engine version that includes the change to Dart. It looks like any "green" engine build after https://github.com/flutter/engine/commit/5a730c60d3cee8deb6901d9e14ae9971ec54b720 would do that.

**Comment by flar** [comment]
It looks like https://github.com/flutter/engine/commit/5a730c60d3cee8deb6901d9e14ae9971ec54b720 rolled into Engine cleanly so using that specific hash should work to get a clean roll into Framework.

**Comment by flar** [comment]
All remaining issues fixed with the following PRs:

https://github.com/flutter/flutter/pull/47896
https://github.com/flutter/flutter/pull/47899

**Comment by yuanhuihui** [comment]
> All remaining issues fixed with the following PRs:
> 
> #47896
> #47899

In my own engine, we have fix flutter_gallery.  Because of using a 7-day old engine, this timeline_summary PR failed to merge before.  So I plan to submit a fix flutter_gallery PR after the timeline_summary PR have merged. timeline_summary.  It seems that you have discovered and solved this problem

**Pull request merged**
